### DISCUSSION TITLE
CodeQL with submodules

### DISCUSSION THREAD

**Discussion by skliper** [community/community#21632] [New to GitHub]
<p>Looking for best practices on setting up CodeQL analysis with submodules.  Currently we are running the analysis from the top repo, but for any alerts in the submodules it reports “Preview unavailable”, “Sorry, we couldn’t find this file in the repository.”</p>
<p>Is there a way to tell CodeQL to look in the submodule repo?</p>
<p>Also we’d really like to run the analysis on submodule pull requests, is there a preferred method?  My thought is to checkout the top repo recursively, then checkout the pull request reference and run the analysis and hopefully the report will then actually show the code with the alert?  Any better way to tell CodeQL we really only care about alerts from the specific submodule?</p>
<p>EDIT - I did just find the configuration capability to specify directories… so one part solved</p>

**Comment by skliper** [comment]
<p>Ok, so I’m checking out the main repo with submodules (gets default submodule), then checking out the submodule (gets the reference submodule) into the expected directory… so I think I’m in business, and should be able to build the code and run analysis on just the submodule (using the path option)</p>

**Comment by criemen** [comment]
<p>I’m happy to hear that! Let us know if you need help with anything.</p>

**Comment by skliper** [comment]
<p>The only remaining issue is in reviewing code scanning alerts I still get “Preview unavailable Sorry, we couldn’t find this file in the repository.” Since the submodule is down 1 directory (builds from main_repo_dir):</p>
<p>main_repo_dir<br>
|_ submodule_repo_dir</p>
<p>So all code references look like submodule_repo_dir/code.c#L52, but to resolve from within the submodule it would need to be code.c#L52 (without the submodule_repo_dir).  Is there any way to strip leading directory from the reported reference?  It’s not all that hard to figure out things by hand, but it would be nice if there was a way to make the links work.  The simple example here isn’t that bad, but imagine 10+ submodules all at different locations in the directory structure…</p>

**Comment by criemen** [comment]
<p>Hi, I checked in with the engineering team for this, and the missing context for annotations within submodules is a known limitation.</p>
<p>However, stripping the directory alone would not be enough here, as we (at that point) do not look through the submodule reference at all.<br>
This requires more work from our side, and we’re tracking that issue internally.</p>

**Comment by skliper** [comment]
<p>Thanks for the feedback, looking forward to improvements!  Running CodeQL analysis in our workflow has been very beneficial!</p>
<p>I’m actually trying it two ways, one is running the analysis as an action from the top repo (which makes sense that it’s extra work to get to the submodule reference), but I’m also running the analysis from the submodule to analyze PRs at that level.  I’d think when run from the submodule the path just needs to be correct?  Example of the submodule analysis at <a href="https://github.com/skliper/osal/blob/4de3def9cd60a72486616f682800449d2953dbfe/.github/workflows/codeql-build.yml" class="inline-onebox" rel="noopener nofollow ugc">osal/codeql-build.yml at 4de3def9cd60a72486616f682800449d2953dbfe · skliper/osal · GitHub</a> (nasa/cFS is the main repo, osal is the submodule of interest).</p>

**Comment by criemen** [comment]
<p>Hi,</p>
<p>I agree that the second way should work.<br>
I can’t see any runs of your codeql action in <a href="https://github.com/skliper/osal/actions" class="inline-onebox">Actions · skliper/osal · GitHub</a> though?</p>
<p>I would expect that if you run your workflow on the osal repo, it should do an analysis of that, and also report all findings with preview et al.</p>

**Comment by skliper** [comment]
<p>Thanks for looking!  Just updated the workflow to run on all pushes.  See <a href="https://github.com/skliper/osal/actions" class="inline-onebox" rel="noopener nofollow ugc">Actions · skliper/osal · GitHub</a>, and the reference issue in <a href="https://github.com/skliper/osal/security/code-scanning/93?query=ref%3Arefs%2Fheads%2Ffix775-add_codeql" rel="noopener nofollow ugc">https://github.com/skliper/osal/security/code-scanning/93?query=ref%3Arefs%2Fheads%2Ffix775-add_codeql</a> (I think the osal/src/tests/time-base-api-test/ <strong>time-base-api-test.c</strong> <strong><span class="hashtag">#L223</span></strong> link just needs “osal/” removed"</p>

**Comment by criemen** [comment]
<p>Sorry for the really late reply, somehow this thread somehow slipped through the cracks.</p>
<p>I am in conversations now with the engineering teams - two workarounds I had in my mind already don’t work, I’ll get back to you if I find something that does.</p>

**Comment by criemen** [comment]
<p>a) We are tracking a bug in the <code>codeql-action</code> that we expose a setting that you could use to set the source root of the analysis to <code>osal</code>, that would fix all your problems.<br>
Unfortunately, there’s no ETA for that.<br>
b) In the meantime, you can do the following:<br>
Under the hood, the codeql action uses the codeql CLI.<br>
If you use the codeql CLI directly (instead of the action) you have access to the relevant setting.<br>
You need to (approximately) do the following:</p>
<ol>
<li>Run your setup steps for the build</li>
<li>Run <code>codeql database create</code> with flags for cpp, and (that’s the thing the action can’t do) <code>--source-root=osal</code>. As build command, put in your <code>make</code> invocation that builds all targets you’re interested in.</li>
<li>Run <code>codeql database analyze</code> on the newly created database from step two, with the suites you are interested in. Use <code>--format=sarif-latest</code> to get a sarif file out of that.</li>
<li>Run <code>codeql github upload-results</code> with that sarif file to get the analysis results into code scanning. The values for the required parameters should all be supplied by the various Actions env vars.</li>
</ol>
<p>See also <a href="https://github.blog/changelog/2021-03-12-codeql-code-scanning-improvements-for-users-analyzing-codebases-on-3rd-party-ci-cd-systems/" class="inline-onebox">CodeQL Code Scanning: improvements for users analyzing codebases on 3rd party CI/CD systems - GitHub Changelog</a> for this workflow. That blog post also links to the codeql CLI documentation.<br>
In general, I find that <code>codeql &lt;subcommand&gt; --help</code> is also of high quality.</p>
<p>I believe that you will not need to install <code>codeql</code> on the actions workers, as that should come pre-installed.</p>

**Comment by ArielSAdamsNASA** [comment]
<aside class="quote group-github-employees" data-username="criemen" data-post="10" data-topic="159839">
<div class="title">
<div class="quote-controls"></div>
 criemen:</div>
<blockquote>
<p>I believe that you will not need to install <code>codeql</code> on the actions workers, as that should come pre-installed.</p>
</blockquote>
</aside>
<p>In theory, to use the codeql commands in a GitHub actions workflow, codeql-actions will need to be initialized using <code>github/codeql-action/init@v1</code>? I tried this <a href="https://github.com/ArielSAdamsNASA/osal/actions/runs/799700494/workflow" rel="noopener nofollow ugc">here</a>, and was given an error that the codeql command was not found.</p>
<p>I was also looking into possibly downloading the zip CodeQL Bundle, but I am having issues doing so on the command line.</p>

**Comment by criemen** [comment]
<p>Hi <a class="mention" href="/arielsadamsnasa">@ArielSAdamsNASA</a> ,<br>
Using <code>github/codeql-action/init</code> will run some commands in the background that will clash with using the CLI directly.<br>
You can see <a href="https://github.com/actions/virtual-environments/blob/365e5a01806196fea730003ea06b56ef145ed79e/images/linux/scripts/installers/codeql-bundle.sh#L14">here</a> where the codeql-bundle will be installed. If you add that directory (or its <code>codeql</code> subdirectory) to the PATH, you should be good to go.<br>
The env variable <code>AGENT_TOOLSDIRECTORY</code> is available during job execution.</p>
<p>Otherwise, you should be able to install the latest CLI release from <a href="https://github.com/github/codeql-cli-binaries/releases/latest">https://github.com/github/codeql-cli-binaries/releases/latest</a> and use that. Follow the instructions at <a href="https://codeql.github.com/docs/codeql-cli/getting-started-with-the-codeql-cli/to">https://codeql.github.com/docs/codeql-cli/getting-started-with-the-codeql-cli/to</a> also get the necessary queries.</p>
<p>Let me know if you need any more help!</p>

**Comment by ArielSAdamsNASA** [comment]
<p>Thank you! I was able to use the latest CLI release in a GitHub Actions workflow, however, the <a href="https://github.com/ArielSAdamsNASA/osal/security/code-scanning/4682?query=ref%3Arefs%2Fheads%2FCodeQLPreview" rel="noopener nofollow ugc">preview</a> is still unavailable.</p>

**Comment by criemen** [comment]
<p>Can you take a screenshot? This page is not publicly accessible.</p>

**Comment by ArielSAdamsNASA** [comment]
<p>Workflow:</p>
<pre><code class="lang-auto">name: "CodeQL Analysis"

on:
  push:
  pull_request:

env:
  SIMULATION: native
  ENABLE_UNIT_TESTS: true
  OMIT_DEPRECATED: true
  BUILDTYPE: release

jobs:

  CodeQL-Build:
    runs-on: ubuntu-18.04
    timeout-minutes: 15
    
    steps:
    
      - name: Checkout bundle
        uses: actions/checkout@v2
        with:
          repository: nasa/cFS
          submodules: true

      - name: Checkout submodule
        uses: actions/checkout@v2
        with:
          path: osal 

      - name: Check versions
        run: git submodule

      - name: Check versions
        run: git submodule

      # Setup the build system
      - name: Set up for build
        run: |
          cp ./cfe/cmake/Makefile.sample Makefile
          cp -r ./cfe/cmake/sample_defs sample_defs
          make prep
   
      - name: CodeQL Download
        run: |
            mkdir codeql
            cd codeql &amp;&amp; wget https://github.com/github/codeql-action/releases/download/codeql-bundle-20210421/codeql-bundle-linux64.tar.gz
            tar xvzf codeql-bundle-linux64.tar.gz codeql   
            readlink -f codeql
            
      #Run codeql database create
      - name: CodeQL Database
        run: |
            cd ../
            ./osal/codeql/codeql/codeql database create osal-database --language=cpp --source-root=osal --command='make -j native/default_cpu1/osal/'
            
      #Run codeql database analyze on the newly created database from step two
      - name: CodeQL Analyze 
        run: |
            cd ../
            ./osal/codeql/codeql/codeql database analyze osal-database /home/runner/work/osal/osal/codeql/codeql/qlpacks/codeql-cpp/JPL_C/ --format=sarif-latest --output=analysis.SARIF

        
      #Run codeql github upload-results with that sarif file 
      - name: CodeQL Upload
        run: |
            cd ../
            ./osal/codeql/codeql/codeql github upload-results --sarif=analysis.SARIF --ref=refs/heads/CodeQLPreview --repository=ArielSAdamsNASA/osal --commit=$GITHUB_SHA --github-auth-stdin=${{ secrets.GITHUB_TOKEN }}
</code></pre>
<p>Output:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/5/d/5d2b81b61a50ab78dd6ef952352ba5f394424f86.png" data-download-href="/uploads/short-url/didChglIOTVGrwDtRmAWVxByHmC.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/69638935/181065920-04b6f6aa-7840-4ed3-baf7-94bf5f1f6f72.png" alt="image" data-base62-sha1="didChglIOTVGrwDtRmAWVxByHmC" width="690" height="356" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/5/d/5d2b81b61a50ab78dd6ef952352ba5f394424f86_2_690x356.png, https://aws1.discourse-cdn.com/github/optimized/3X/5/d/5d2b81b61a50ab78dd6ef952352ba5f394424f86_2_1035x534.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/5/d/5d2b81b61a50ab78dd6ef952352ba5f394424f86_2_1380x712.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/5/d/5d2b81b61a50ab78dd6ef952352ba5f394424f86_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1643×850 48.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by criemen** [comment]
<p>That is strange! I’ve pinged the engineering team, and will get back to you once I’ve heard from them.</p>

**Comment by criemen** [comment, marked as answer]
<p>Hi Ariel,</p>
<p>the engineering team confirmed this behavior as normal.<br>
The reason is that we currently don’t look for files across submodules for the preview.<br>
There is an internal feature request that we produce the snippet that’s shown there as part of the analysis of the code, and then use that for the preview window, but that feature is not being worked on right now.</p>
<p>Let me know if I can help you further.</p>
